compiler/luci/pass/src/RemoveUnnecessaryTransposeNetPass.cpp

```diff
@@ -0,0 +1,483 @@
+/*
+ * Copyright (c) 2023 Samsung Electronics Co., Ltd. All Rights Reserved
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "luci/Pass/RemoveUnnecessaryTransposeNetPass.h"
+
+#include <luci/IR/CircleNodes.h>
+#include <luci/Profile/CircleNodeOrigin.h>
+
+#include <vector>
+
+namespace
+{
+
+class TaggedShapeAnalyzer final
+{
+public:
+  /**
+   * @brief check 'Transpose-Reshape-Transpose' can be replaced by one 'Reshape'.
+   *
+   * @example
+   *  Let's explain how analyzer check Transpose-Reshape-Transpose pattern with an exact example.
+   *
+   *  Let's assume under pattern is given :
+   *
+   *      Input(1, 7, 7, 448)
+   *            |
+   *      Transpose(perm=(0, 3, 1, 2))
+   *            |
+   *      Resahape(shape=(1, 448, 49))
+   *            |
+   *      Transpose(perm=(0, 2, 1))
+   *            |
+   *      Output(1, 49, 448)
+   *
+   *
+   *  It simulates how dimensions of the tensor's shape are transformed/moved using a member
+   *  variable named '_shape'.
+   *  'tags' in '_shape' record the initial order of each dimension.
+   *
+   *   TIMELIE               |   _shape states :
+   *                         |
+   *   init_shape            |    _shape :
+   *                         |        - value :   (1)   (7)   (7)   (448)
+   *                         |        - tags  :   (-)   (0)   (1)   (2)
+   *                         |
+   *   analyze_transpose     |    _shape :
+   *                         |        - value :   (1)   (448)  (7)  (7)
+   *                         |        - tags  :   (-)   (2)    (0)  (1)
+   *                         |
+   *   analyze_reshape       |    _shape :
+   *                         |        - value :   (1)   (448)   (49)
+   *                         |        - tags  :   (-)   (2)     (0, 1)
+   *                         |
+   *   anaylze_transpose     |    _shape :
+   *                         |        -value  :   (1)   (49)    (448)
+   *                         |        -tags   :   (-)   (0, 1)  (2)
+   *
+   *  If final state of tags are in incremental-order, it gurantees that the pattern doesn't change
```

Comment:
I don't get this... -_-;

Reply:
Thank you a lot for your time. I hope it is in progress to get better. :smile:  

May I ask you which point is hard to understand? 
The total example I tried to explain with the `_shape` state?  Or the last sentence start with `If final state of tags are ~ ..` ?

Reply:
> hich point is hard to understand?

`in incremental-order,` suddenly becomes `gurantees that the pattern doesn't change`
I feel there is some logical jump.